feat: add dynamic mDNS interface management with runtime monitoring

randomdudebunchofnumbers opened the pull request:
## Summary

Implements dynamic mDNS interface management that gracefully handles network interface availability changes at runtime. Fixes #62 

**Note**: While the primary fix targets configurations with specific interfaces, the monitoring infrastructure benefits all deployment modes by tracking interface state changes.

## Changes

**Core Implementation:**
- Modified `interfaces()` method to handle partial interface availability (non-fatal resolution)
- Added continuous interface monitoring with 15-second polling interval (when specific interfaces configured)
- Implemented automatic re-announcement when interfaces appear/disappear
- Added proper cleanup on shutdown to prevent goroutine leaks

**Code Quality:**
- Added `getUsableInterface()` helper function to reduce code duplication
- Implemented thread-safe state tracking with proper mutex protection
- Added comprehensive unit tests for new functionality

## Key Features

✅ **Graceful degradation**: Announces on available interfaces even when some are missing  
✅ **Runtime adaptation**: Automatically detects and responds to interface changes (appear/disappear/re-appear)  
✅ **Backwards compatible**: No API changes; existing code continues to work unchanged  
✅ **Clean shutdown**: Proper goroutine lifecycle management prevents resource leaks  
✅ **Test coverage**: Added tests for partial interface availability and helper functions

## Behavior

### When specific interfaces are configured:
- ✅ Announces on available interfaces, even if some are missing
- ✅ Monitors interface state every 15 seconds
- ✅ Re-announces automatically when interfaces appear or disappear

### When no specific interfaces are configured (default):
- ⚠️ **Known limitation**: Still announces on the interface list captured at startup
- ⚠️ Does not yet adapt to runtime interface changes
- 🔄 **Future work**: Extend monitoring to detect when new interfaces appear/existing ones disappear

Commit:
feat: add dynamic mDNS interface management with runtime monitoring

Commit:
improved test coverage

Commit:
changed interface names used for testing to avoid name collisions with actually existing interfaces

randomdudebunchofnumbers commented:
I added another commit to fix the broken unit tests. The problem was that two tests used real interface names ("eth0", "eth1") that may exist on CI systems like GitHub Actions runners, causing test failures.

DerAndereAndi commented on mdns/mdns.go:
```diff
@@ -156,28 +164,80 @@ func (m *MdnsManager) interfaces() ([]net.Interface, []int32, error) {
+	}
+	if m.currentIfaces == nil {
+		m.currentIfaces = make([]string, 0)
+	}
```
Shouldn't these both be reset all time? Otherwise the same interface could be added multiple times.

When an interface appears:
1. attemptResolveMapping() sets currentIfaces = ["eth0"]
2. Then calls reannounceWithNewInterfaces() → interfaces()
3. interfaces() sees currentIfaces is not nil, skips initialization
4. Appends again: currentIfaces = ["eth0", "eth0"] - duplicate

DerAndereAndi commented on mdns/mdns.go:
```diff
@@ -239,14 +304,181 @@ func (m *MdnsManager) Start(cb api.MdnsReportInterface) error {
+	}
+}
+
+// updateProviderInterfaces updates the provider's interface list
+func (m *MdnsManager) updateProviderInterfaces(ifaces []net.Interface, ifaceIndexes []int32) {
+	// Update provider's interface list
+	switch provider := m.mdnsProvider.(type) {
+	case *AvahiProvider:
+		provider.ifaceIndexes = ifaceIndexes
+	case *ZeroconfProvider:
+		provider.ifaces = ifaces
+	}
+}
```
Directly writing to the provide fields can lead to data races.

For `ZeroconfProvider`, `ifaces` is used in `chanListener()` at zeroconf.go:89:
```
go func() {
    _ = zeroconf.Browse(z.ctx, ..., zeroconf.SelectIfaces(z.ifaces))  // Read
}()
```

Impact: Data race if updateProviderInterfaces() is called while a browse operation is using ifaces.

DerAndereAndi commented on mdns/mdns.go:
```diff
@@ -239,14 +304,181 @@ func (m *MdnsManager) Start(cb api.MdnsReportInterface) error {
+// updateProviderInterfaces updates the provider's interface list
+func (m *MdnsManager) updateProviderInterfaces(ifaces []net.Interface, ifaceIndexes []int32) {
+	// Update provider's interface list
+	switch provider := m.mdnsProvider.(type) {
```
if `m.mdnsProvider` is nil (e.g. during shutdown), this could cause issues. we should check for nil before the switch.

DerAndereAndi commented on mdns/mdns.go:
```diff
@@ -239,14 +304,181 @@ func (m *MdnsManager) Start(cb api.MdnsReportInterface) error {
+	}
+
+	m.refreshStopChan = make(chan struct{})
+	m.refreshTicker = time.NewTicker(15 * time.Second)
```
The interval should be a constant that's defined in the top of this file.

DerAndereAndi reviewed (changes requested):
Thanks for the pull request, overall looks good! I added some comments, please check them out.

Commit:
Fix mDNS interface management issues from code review

- Reset interface trackers on each call to prevent duplicates
- Add thread-safe setters/getters for provider interface fields
- Add nil check in updateProviderInterfaces() for shutdown safety
- Extract refresh interval to named constant

Commit:
Fix race condition in mDNS interface reappearance detection

Commit:
Refactor: deduplicate interface resolution logic and test helpers

Commit:
Fix mDNS reannouncement disrupting connections on unrelated interfaces

Commit:
Merge remote-tracking branch 'origin/dev' into feat/dynamicMdnsInterfaceManagement

Commit:
fix merge error

randomdudebunchofnumbers commented:
Hello,

I updated my pull request and kindly ask you for another review. However, there is one known limitation:

When interfaces change, both the zeroconf and avahi providers must tear down and recreate their announcements with the updated interface list. During this teardown, goodbye packets (TTL=0) are sent for the old records. Even though the new announcement is created before the old one is shut down, remote devices may still process the goodbye packets and briefly consider the service as having left the network — which is the exact scenario the create-then-swap pattern was intended to prevent.

For zeroconf, this is unavoidable without extending the library, as zeroconf.Register() creates an immutable server bound to a fixed interface list.

For avahi, this could be solved by maintaining a separate entry group per interface, allowing individual interfaces to be added or removed without affecting announcements on other interfaces. This is not yet implemented.

DerAndereAndi commented on mdns/mdns.go:
```diff
@@ -153,31 +165,100 @@ func NewMDNS(
+		return ifaces, ifaceIndexes, nil
+	}
+
+	// Reset and rebuild tracking state
```
Don't we need to use locking on refreshMux when writing the following 2 fields?

randomdudebunchofnumbers replied:
`interfaces()` is currently only called from `Start()` before the refresh goroutine is spawned, so there's no race in practice. But you're right that the locking discipline should be consistent. I'll add refreshMux protection around the writes to missingIfaces and currentIfaces in `interfaces()` to match how `attemptResolveMapping()` guards them.

DerAndereAndi commented on mdns/avahi.go:
```diff
@@ -211,8 +229,16 @@ func (a *AvahiProvider) Announce(serviceName string, port int, txt []string) err
+	oldEntryGroup := a.avEntryGroup
 	a.avEntryGroup = entryGroup
 
+	if oldEntryGroup != nil {
```
We might want to release the mux lock before this if statement instead of using `defer a.mux.Unlock()`. If the `.EntryGroupFree` call is blocked, e.g. due to dBUS, then the mutex lock blocks accessing these fields.

randomdudebunchofnumbers replied:
I'll update `Announce` and `Unannounce` to release the mutex before calling `EntryGroupFree`, so a blocked dBUS call won't hold the lock. 

Side note: calling `EntryGroupFree`after releasing the mutex is actually fine, since avahi holds its own internal mutex.

DerAndereAndi commented on mdns/mdns.go:
```diff
@@ -153,31 +165,100 @@ func NewMDNS(
 	if len(ifaces) == 0 {
-		ifaces = nil
-		ifaceIndexes = []int32{avahi.InterfaceUnspec}
+		return nil, nil, nil
```
We are not returning any data, so we might want to return an error here?

```
var ErrNoInterfacesAvailable = errors.New("none of the configured interfaces are available")
```

Then in here:
```
if len(ifaces) == 0 {
    return nil, nil, ErrNoInterfacesAvailable
}
```

And in reannounceWithNewInterfaces():
```
ifaces, ifaceIndexes, err := m.resolveInterfaces()
if err != nil {
    if errors.Is(err, ErrNoInterfacesAvailable) {
        logging.Log().Debug("mdns: still no interfaces available during refresh")
        if wasAnnounced {
            m.UnannounceMdnsEntry()
        }
        return
    }
    // handle unexpected errors
}
```

This makes the "no interfaces" case explicit and prevents future callers from accidentally treating nil, nil, nil as success.

randomdudebunchofnumbers replied:
Will implement it as suggested.

DerAndereAndi commented on mdns/mdns.go:
```diff
@@ -239,14 +325,188 @@ func (m *MdnsManager) Start(cb api.MdnsReportInterface) error {
+		return
+	}
+
+	// Update provider's interface list using thread-safe setters
```
How about adding an interface method to each provider like

```
UpdateInterfaces(ifaces []net.Interface, ifaceIndexes []int32)
```

Then here we can simply call:
```
if updater, ok := m.mdnsProvider.(InterfaceUpdater); ok {
    updater.UpdateInterfaces(ifaces, ifaceIndexes)
}
```

The removes the type check, any new implementation won't be missed here and the implementation will then use either `ifaces` or `ifaceIndexes`

randomdudebunchofnumbers replied:
Thanks for the suggestion! I agree this would be a cleaner approach. Would it be okay to handle this in a follow-up PR to keep the scope of this one focused?

DerAndereAndi commented on mdns/mdns.go:
```diff
@@ -153,31 +165,100 @@ func NewMDNS(
+		return nil, nil, nil
 	}
 
+	logging.Log().Infof("mdns: using %d of %d required interfaces", len(ifaces), len(m.ifaces))
```
This fires on every re-announcement triggered by `attemptResolveMapping`. Since `attemptResolveMapping` already logs the specific appeared/disappeared interfaces at Info level, this additional Info log is redundant for the change-triggered path. It is useful for the initial `Start()` path though.

randomdudebunchofnumbers replied:
I'll remove the redundant log. Thank you for your time and your feedback. I'll push my changes shortly.

DerAndereAndi reviewed (changes requested):
Great changes, thanks already. I got some more, what do you think?

Commit:
added refreshMux protection around the writes to missingIfaces and currentIfaces in interfaces()

Commit:
release the mutex before calling EntryGroupFree

Commit:
added additional unit test covering refreshMutex protection

Commit:
return ErrNoInterfacesAvailable instead of nil when no configured interfaces are usable

Commit:
removed redundant log

Commit:
merged latest dev

randomdudebunchofnumbers commented:
I updated my feature branch with the latest from dev. It'd be great if you could take a look. Thank you very much.

DerAndereAndi committed:
[mDNS] Add reproducer tests for dynamic interface management issues

Add failing tests that expose concurrency and lifecycle bugs in the
dynamic mDNS interface monitoring introduced in PR #61:

- Shutdown does not wait for refresh goroutine to exit (nil deref risk)
- mdnsProvider read/write race between refresh goroutine and Shutdown
- updateProviderInterfaces silently skips non-concrete provider types
- isAnnounced not reset when Unannounce panics during Shutdown

Also adds a regression guard for Avahi Announce mutex release on all
error paths (currently correct).

DerAndereAndi committed:
[mDNS] Fix refresh goroutine lifecycle, provider update, and stale state

- stopInterfaceRefresh now waits for refreshLoop goroutine to exit
  before returning, preventing Shutdown from setting mdnsProvider=nil
  while the goroutine is still accessing it (fixes race + nil deref)
- Replace type-switch in updateProviderInterfaces with a single
  UpdateInterfaces interface so any provider gets updated, not just
  concrete Avahi/Zeroconf types
- Reset isAnnounced at the end of Shutdown to prevent stale state
  when UnannounceMdnsEntry panics inside the recover wrapper

DerAndereAndi commented:
Thanks @randomdudebunchofnumbers ! I added some more fixes that came up and changed the approach to the interface pattern.

DerAndereAndi committed:
[mDNS] Fix avahi Shutdown deadlock when chanListener is busy

AvahiProvider.Shutdown() sent on an unbuffered shutdownChan while holding
a.mux. If chanListener was not currently sitting at its select -- e.g.
because it had received a service from addServiceChan and was inside
processService -- the send blocked forever: Shutdown could not release
a.mux because the send was pending, and chanListener could not return
to the select to receive it because it was busy elsewhere.

The dynamic-interface PR made this materially worse by introducing
getIfaceIndexes() inside processService, which acquires a.mux. Previously
processService read a.ifaceIndexes without locking, so at worst Shutdown
had to wait for an in-flight ResolveService. With the new code path,
Shutdown holding a.mux actively blocks chanListener inside
getIfaceIndexes, guaranteeing the cycle.

Fix:
- Buffer shutdownChan with capacity 1 so Shutdown's signal never blocks,
  regardless of chanListener's state.
- Pass shutdownChan/addServiceChan/removeServiceChan as arguments to
  chanListener so it no longer reads the provider fields directly. Before
  the buffer change, the unbuffered send was accidentally serving as a
  happens-before barrier between chanListener's field reads and
  Shutdown's subsequent nil assignments. Buffering removed that barrier,
  which the race detector immediately caught in Test_Avahi,
  Test_Avahi_Reconnect and Test_chanListener. Capturing the channels at
  goroutine spawn time follows the same pattern already used by
  refreshLoop in mdns.go.

Adds Test_AvahiShutdownDoesNotDeadlockOnBusyChanListener which
deterministically pins chanListener inside processService via a blocking
ResolveService mock and asserts that Shutdown returns within 500ms.
Fails against the pre-fix code and passes with both changes applied.

Verified with: go test -race ./... and
go test -tags=deadlock -race ./mdns/... ./ship/... ./hub/...

sthelen-enqs reviewed (approved):
LGTM, thanks
